File: problems/leetcode/offer24.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if head == None:
            return None
        p, q = None, head
        r = q.next
        while q != None:
            q.next = p
            p = q
            q = r
            if r != None:
                r = r.next
        return p

File: problems/leetcode/offer24.cpp
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head) return nullptr;
        ListNode *p = nullptr, *q = head, *r = q -> next;
        while(q) {
            q -> next = p;
            p = q;
            q = r;
            if (r) r = r -> next;
        }
        return p;
    }
};